### ISSUE TITLE
AWS lambda instrumentation fails to propagate HTTP headers context

### ISSUE THREAD

**Issue by kubawach** [open-telemetry/opentelemetry-java-instrumentation#2131]
**Describe the bug**
- OTEL lambda tracer (io.opentelemetry.instrumentation.awslambda.v1_0.AwsLambdaTracer) sets parent span using either XRAY header or HTTP headers
- XRAY is preferred if header (env value - _X_AMZN_TRACE_ID) is set
- HTTP is used if XRAY parent is invalid (or XRAY header was not set)
- Currently AWS sets XRAY header not matter lambda settings - ie even with XRAY disabled in the console
- XRAY disabled - value is eg `Root=1-60115e2e-3bc996f80f14f6e33d18f37b;Parent=4761858a3c680442;Sampled=0` (notice sampled settings)
- XRAY enabled - value is eg `Root=1-60115e2e-3bc996f80f14f6e33d18f37b;Parent=4761858a3c680442;Sampled=1` (notice sampled settings)

**Steps to reproduce**
- deploy lambda with OTEL wrapper, configure any exporter

**What did you expect to see?**
- HTTP headers should be used (if set) once XRAY is disabled

**What did you see instead?**
- XRAY propagation in all cases (enabled, disabled)

**What version are you using?**
- latest (0.14.0, master)

**Environment**
NA

**Additional context**

1. First question is whether AWS should set XRAY header when tracing is disabled? If it didn't (which I think was the case some time ago), everything would work as expected, that is:
- XRAY enabled - propagation of XRAY context
- XRAY disabled, HTTP headers - propagation of HTTP context
- XRAY disabled, no HTTP headers - no propagation (no parent span)

2. If above behaviour can't be changed, I believe that the right way is to set HTTP propagation as preferred, this way we will get slightly different but still acceptable (in my opinion of course! :) ) behaviour: 
- XRAY enabled, HTTP headers - propagation of HTTP context
- XRAY enabled,  no HTTP headers - propagation of XRAY context
- XRAY disabled, HTTP headers - propagation of HTTP context
- XRAY disabled, no HTTP headers - propagation of XRAY context with parent that should not be sampled

@anuraaga please have an "expert look" here

**Comment by anuraaga** [comment]
Hmm - this might be some strange behavior I've seen before where if you don't ever enable x-ray, lambda doesn't have traces. But if you enable it once, and then disable it later, it still creates traces and that environment variable.

> XRAY enabled, HTTP headers - propagation of HTTP context

This doesn't work unfortunately since we need to parent to the lambda context, which might be a child of an API gateway span, etc since these are X-Ray users.

Since user has to enable xray propagator, does this sort of logic work?

```java
if (noEnvVar) {
  return extract(httpHeaders);
} else {
  Context c = extract(envVar);
  if (c.isValid()) {
    return c;
  }
  return extract(httpHeaders);
```

**Comment by kubawach** [comment]
I think I might have not expressed myself clearly enough :) What I meant was that:
1. `io.opentelemetry.instrumentation.awslambda.v1_0.AwsLambdaTracer#parent` method always first checks for parent from XRay header - therefore as header (env var in fact) is always present (even if XRay is disabled for a function), the parent will always be from XRay (even if user enabled HTTP propagator and sent appropriate headers)
2. `XRAY enabled, HTTP headers - propagation of HTTP context` - meaning "XRay enabled for a function, XRay parent sampled = 1, HTTP propagation headers also present and valid"

I believe that above ^^ could be acceptable, that is if user wanted to use XRay instead of HTTP-headers tracing (with some other backend ?) he'd simply need to configure lambda without HTTP propagators.

As always, let me know what you think

**Comment by anuraaga** [comment]
@kubawach Let's use the term lambda instead of XRay here for parenting, I think that may be confusing is.

> io.opentelemetry.instrumentation.awslambda.v1_0.AwsLambdaTracer#parent method always first checks for parent from XRay header - therefore as header (env var in fact) is always present (even if XRay is disabled for a function), the parent will always be from XRay (even if user enabled HTTP propagator and sent appropriate headers)

IIUC, you are saying to always read the `X-Amzn-Trace-Id` header (or any other header) first. The problem with this is upstream, for example the user themselves, may have sent `X-Amzn-Trace-Id` as a header to API gateway. API gateway's parent becomes this upstream node. Lambda's parent is supposed to be API gateway, but the HTTP header will still have that upstream span ID, the environment variable has a different span ID corresponding to API gateway, so we need to read from the variable.

So my proposal is to always try reading from the environment variable first - if user has not enabled XRay propagator, that will never return a valid spancontext, and we then check out the headers. How does that sound?

**Comment by kubawach** [comment]
@anuraaga I completely agree with above reasoning, it was working great, until recently :) I discovered that even if XRay is disabled for a lambda, the environment variable "_X_AMZN_TRACE_ID" still contains a valid spancontext, but with sampling value set to NO (0). Also, current code always tries to extract from env var using AwsXrayPropagator (this is enforced, user does not need to enable it). 
References:
- choosing parent (env variable priority): io.opentelemetry.instrumentation.awslambda.v1_0.AwsLambdaTracer#parent
- enforcing AwsXRayPropagator to extract the value: io.opentelemetry.instrumentation.awslambda.v1_0.ParentContextExtractor#fromXRayHeader

To sum up, for HTTP (Api Gateway Proxy) events, user can't set the parent with HTTP headers (eg using OTel instrumented HTTP client). 

So I still see two options only:
- AWS stops setting the variable if XRay is disabled 
- we change the code so that we first check for "configured propagators HTTP headers" and if not found we try "_X_AMZN_TRACE_ID" env variable

**Comment by anuraaga** [comment]
Oh it's `a valid spancontext, but with sampling value set to NO (0)`? Can we fallback to HTTP headers after reading the variable and seeing sampling flag is disabled? And yeah I missed that, I think it's fine to remove the `AwsXrayPropagator` hard reference if this can work.

Otherwise I think we need to make it flag-configurable. I may still be misunderstanding, but `we change the code so that we first check for "configured propagators HTTP headers" and if not found we try "_X_AMZN_TRACE_ID" env variable` doesn't work for API Gateway + X-Ray right, since the HTTP header context will win?

**Comment by kubawach** [comment]
The first options sounds very reasonable (only minor change of the code required). The only risk is that XRay parent will never be taken into account if set to `sampled = false`. I don't see currently a use case where this would be a *real* issue tho :)

The second option would allow API Gateway + X-Ray to work if caller didn't provide any HTTP propagation headers explicitly. That is if lambda was called not by using an instrumented HTTP client with configured propagation. The decision would be taken only based on HTTP headers (fields) of propagators configured for a lambda yielding a valid spancontext.

**Comment by kubawach** [comment]
@anuraaga please let me know what would you prefer, disregarding "sampled=false" parent from "_X_AMZN_TRACE_ID" env variable (option 1 as discussed) or giving priority to HTTP headers derived parent (option 2 as discussed). Thanks!

**Comment by anuraaga** [comment]
Option 1 seems good to me.

**Comment by kubawach** [comment]
Great, will proceed accordingly :)
